src/rust/engine/fs/store/src/lib.rs

```diff
@@ -1270,7 +1297,22 @@ impl Store {
                   Permissions::Writable if f.is_executable() => 0o755,
                   Permissions::Writable => 0o644,
                 };
-                store.materialize_file(path, f.digest(), mode).await
+
+                if immutable_inputs.is_some()
+                  && f.digest().size_bytes > IMMUTABLE_FILE_SIZE_LIMIT
+                  && !mutable_paths.contains(&path)
+                  && !mutable_paths.iter().any(|p| path.starts_with(p))
```

Comment:
This is me being very lazy and trying to avoid `mutable_files` and `mutable_dirs`. `.contains` first for speed (for file-matching), then slower `.any()` to match dirs.

Suggestions welcome!

Reply:
Oh duh, check for contains in the directory branch too, then pass a variable down the recursive call.